File: Stack/Python/Online Stock Span.py
# Problem link: https://leetcode.com/problems/online-stock-span/description/
# Time: O(n)
# Space: O(n)
class StockSpanner:

    def __init__(self):
        self.stack = []

    def next(self, price: int) -> int:
        result = 1

        while self.stack and self.stack[-1][0] <= price:
            result += self.stack.pop()[1]
        
        self.stack.append((price, result))
        return result
        

File: Stack/Java/Online Stock Span.java
// Problem link: https://leetcode.com/problems/online-stock-span/description/
// Time: O(n)
// Space: O(n)
class StockSpanner {

    Stack<int[]> stack;

    public StockSpanner() {
        this.stack = new Stack<>();
    }
    
    public int next(int price) {
        int result = 1;

        while(!stack.isEmpty() && stack.peek()[0] <= price) {
            result += stack.pop()[1];
        }

        stack.push(new int[] {price, result});

        return result;
    }
}
